use PG_MODULE_MAGIC_EXT to report the full version

tureba opened the pull request:
On PG 18+, PG_MODULE_MAGIC_EXT allows the module to report its name and version, where the version can be more fine-grained than the SQL-level extension.

So for an extension installed on a coarser version 1.6:

postgres=# select extname, extversion from pg_extension where extname = 'pg_cron';
 extname | extversion
---------+------------
 pg_cron | 1.6
(1 row)

This change allows the loaded modules to report, not only this...

postgres=# select pg_get_loaded_modules();
 pg_get_loaded_modules
-----------------------
 (,,pg_cron.so)
(1 row)

...but also the bugfix release:

postgres=# select pg_get_loaded_modules();
   pg_get_loaded_modules
----------------------------
 (pg_cron,1.6.7,pg_cron.so)
(1 row)

This change can help troubleshoot cases where the release is relevant and consulting the OS packages is more difficult, or where the .so might have changed without a database server restart.

It does add a small amount of extra work for releases, in the sense that pg_cron.c now has to be updated along with the changelog for every release henceforth.

tureba committed:
use PG_MODULE_MAGIC_EXT to report the full version

On PG 18+, PG_MODULE_MAGIC_EXT allows the module to report its name and
version, where the version can be more fine-grained than the SQL-level
extension.

So for an extension installed on a coarser version 1.6:

postgres=# select extname, extversion from pg_extension where extname = 'pg_cron';
 extname | extversion
---------+------------
 pg_cron | 1.6
(1 row)

This change allows the loaded modules to report, not only this...

postgres=# select pg_get_loaded_modules();
 pg_get_loaded_modules
-----------------------
 (,,pg_cron.so)
(1 row)

...but also the bugfix release:

postgres=# select pg_get_loaded_modules();
   pg_get_loaded_modules
----------------------------
 (pg_cron,1.6.7,pg_cron.so)
(1 row)

This change can help troubleshoot cases where the release is relevant
and consulting the OS packages is more difficult, or where the .so might
have changed without a database server restart.

It does add a small amount of extra work for releases, in the sense that
pg_cron.c now has to be updated along with the changelog for every
release henceforth.

sandeep-edb reviewed (approved):
LGTM

Basha-EDB commented:
@tureba - As the Pull Request has been approved, could you please merge it at your earliest convenience.

tureba committed:
PG_MODULE_MAGIC_EXT: use version from CHANGELOG.md

This causes make to take the latest version from the top of the
CHANGELOG.md file and use it for PG_MODULE_MAGIC_EXT.

This has the potential to reduce the complexity of doing new releases,
by going from three places that the version needs to be synced (git tag,
changelog, pg_cron.c) to two places (git tag and changelog), which
already was the status quo.

It does rely on the assumption that the git tag and the changelog are
kept in sync for every release, which may or may not hold forever.

The alternative would be to get the pg_cron version from git tag, but
that'll be a problem when one is building from a .tar.gz file, which may
be done by distributions nowadays.

tureba commented:
Now I pushed a new commit for the same branch that takes the version from the tip of CHANGELOG.md. I think this is an option. The other is to take the git tag, but I fear the git tag may not be available when distros download the .tar.gz.

marcoslot reviewed (approved):
I think this can be improved  because git describe will just pick up the tag, which is more likely to be accurate, but merging for now. Thanks!
